src/api/c++/z3++.h

```diff
@@ -559,32 +643,30 @@ namespace z3 {
         }
         */
 
-        class iterator {
+        class iterator final {
             ast_vector_tpl const* m_vector;
             unsigned m_index;
         public:
-            iterator(ast_vector_tpl const* v, unsigned i): m_vector(v), m_index(i) {}
-            iterator(iterator const& other): m_vector(other.m_vector), m_index(other.m_index) {}
-            iterator operator=(iterator const& other) { m_vector = other.m_vector; m_index = other.m_index; return *this; }
+            iterator(ast_vector_tpl const* v, unsigned i) noexcept : m_vector(v), m_index(i) {}
 
-            bool operator==(iterator const& other) const {
+            bool operator==(iterator const& other) const noexcept {
```

Comment:
This should never raise an exception, we optimize by marking the function as `noexcept`; this also allows this function/operator to be used safely in `noexcept` contexts